debezium-embedded/src/main/java/io/debezium/embedded/ConnectorEngine.java

```diff
@@ -0,0 +1,1134 @@
+         * stopped. If the connector is not currently running, immediately call the supplied function.
+         *
+         * @param uponCompletion the function to always call when complete, even if the connector is not currently running
+         * @return a future that can be called to wait for the connector to complete their shutdown; never null but empty
+         *         if no connector exists with the given name
+         */
+        public synchronized Future<Boolean> stop(Runnable uponCompletion) {
+            Future<Boolean> result = new WorkerFuture(connectorName, completionLatch, uponCompletion);
+            running.set(false); // signal to task threads to stop
+            return result;
+        }
+
+        /**
+         * Add to the queue of {@link ChangeEvent}s the list of {@link SourceRecord} change events from the given
+         * {@link SourceTask}.
+         * 
+         * @param records the list of source events; may be null or empty
+         * @param task the connector task that generated these events; may not be null
+         * @param sequence a sequence of monotonically increasing long values that can be assigned to each {@link ChangeEvent},
+         *            used to ensure that calling {@link ChangeEvent#complete()} out of order does not lose recorded offsets
+         * @param maxSequenceCommitted the mutable maximum sequence number that has been committed so far
+         * @throws InterruptedException if the calling thread is interrupted while waiting for the events to be enqueued
+         */
+        protected void enqueueRecords(List<SourceRecord> records, SourceTask task, AtomicLong sequence, AtomicLong maxSequenceCommitted)
+                throws InterruptedException {
+            if (records == null) return;
+            int numRecords = records.size();
+            if (numRecords == 0) return;
+            List<ChangeEvent> events = new ArrayList<>(numRecords);
+            for (SourceRecord record : records) {
+                long sequenceNumber = sequence.incrementAndGet();
+                events.add(new SourceChangeEvent(record, () -> commitOffset(record, task, sequenceNumber, maxSequenceCommitted)));
+            }
+            eventQueue.put(events);
+        }
+    }
+
+    protected ConnectorContext workerConnectorContext(String connectorName, ConnectorCallback callback) {
+        return new ConnectorContext() {
+            @Override
+            public void requestTaskReconfiguration() {
+                // Do nothing ...
+            }
+
+            @Override
+            public void raiseError(Exception e) {
+                callback.connectorFailed(connectorName, e.getMessage(), e);
+            }
+        };
+    }
+
+    /**
+     * A {@link Future} implementation that is used to wait for a task to complete.
+     * 
+     * @author Randall Hauch
+     */
+    @Immutable
+    protected static final class WorkerFuture implements Future<Boolean> {
```

Comment:
I may be mistaken, but it seems like this class' behavior could be replaced by the `CompletableFuture` API